### PR TITLE
Extended LUKS encryption support

### PR DESCRIPTION
**Pull request by schaefi**
We supported luks encrypted images with the exception of the
boot partition because grub was not able to read from an
encrypted device at that time. Since some time grub is able
to read from an encrypted partition and this commit add
support for it. With this patch luks encrypted images will
no longer create an extra boot partition and the entire
system will be encrypted. The opening of the luks container
happens on the grub level only once. Customers can still
specify to use an unencrypted extra boot partition with the
bootpartition="true" flag. In this case the opening of the
luks containers happens on the systemd level as before.
This is related to Issue #1162

### CODE REVIEW

File: kiwi/storage/luks_device.py [hunk 1]
```diff
@@ -129,11 +147,39 @@ def create_crypttab(self, filename):
...
+        :param string filename: file path name
+        """
+        min_length = 12
+        max_length = 12
+        character_selection = ''.join(
+            [string.ascii_letters, string.punctuation, string.digits]
+        )
+        passphrase = ''.join(
+            random.choice(character_selection) for x in range(
+                random.randint(min_length, max_length)
```
[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: @dcermak do you have an opinion about the security complain in this case ? From my perspective it's enough security. Thoughts ?

[the remark this answers is not in the document]
Comment (dcermak) [reply kiwi/storage/luks_device.py]: `random.choice` uses a pseudorandom generator and could thus be predictable, so this complaint is actually correct. I would suggest to either use `os.urandom(size)` or `secrets.choice(sequence)` (the latter is only available in Python >= 3.6).

Also, your password generation code is buggy: the generated passphrase is empty.

[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: > Also, your password generation code is buggy: the generated passphrase is empty.

not in my tests and the build image ??

>random.choice uses a pseudorandom generator and could thus be predictable, so this >complaint is actually correct.

ok, I'll adapt to your suggestion with os.urandom

[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: From my build:

```
cat /tmp/mytest/build/image-root/.root.keyfile 
@Hui}iQyc_C"
```

[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: It's never empty

[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: Ok I moved to os.urandom and a 128 byte key length. My test build worked nicely. How about it now ?

File: kiwi/storage/luks_device.py [hunk 2]
```diff
@@ -147,6 +174,26 @@ def is_loop(self):
...
+
+        :param string filename: file path name
+        """
+        min_length = 12
+        max_length = 12
+        character_selection = ''.join(
+            [string.ascii_letters, string.punctuation, string.digits]
+        )
+        passphrase = ''.join(
+            random.choice(character_selection) for x in range(
```
[the remark this answers is not in the document]
Comment (davidcassany) [reply kiwi/storage/luks_device.py]: @schaefi just saw that the `random.choice` is still being used, I though last commits were done to switch to `os.urandom`. Is there something I missed? codacy still complaints.

**Comment by schaefi** [comment]
> Looks good to me, this is a nice feature 
> 
> Off-topic: I am wondering if we could reuse this concept of adding a default keyfile for #874 somehow. A week and half ago I did some tests regarding the use of rd.luks.* dracut and systemd parameters (on tumbleweed with up to date versions) to set a keyfile location to avoid password prompt at boot and I miserably failed. Also it could be investigated if grub supports something like that, then we could just use this setup and forget about rd.luks.* parameters.

If this is merged we would be able to add more keys, also one without a passphrase. But this will invalidate security. I couldn't came across a way to pass in a parameter to grub as input for the master key

File: kiwi/defaults.py [hunk 3]
```diff
@@ -61,6 +61,13 @@ def __init__(self):
             'kiwi_revision'
         ]
 
+    @staticmethod
+    def get_luks_key_length():
+        """
+        Provides key length to use for random luks keys
+        """
+        return 128
```
[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/defaults.py]: ok, I will add some docs about the behavior for encrypted disks

File: kiwi/storage/luks_device.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (schaefi) [reply kiwi/storage/luks_device.py]: I assume with the 256 key length this is ok

[the remark this answers is not in the document]
Comment (dcermak) [reply kiwi/storage/luks_device.py]: Sorry, had a copy+paste fail there, the password is actually never empty. Although the call to `random.randint(min_length, max_length)` was pointless, as `min_length == max_length` and thus it always returns 12. But this code is gone now, so everything is fine.

**Comment by schaefi** [comment]
> This needs some documentation though.

@dcermak I've added a working with images chapter that describes how we support encrypted images. I think it's best to do this based on an example.

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
+.. _setup_for_luks:
+
+KIWI Image Description Encrypted Disk
+=====================================
+
+.. sidebar:: Abstract
+
+   This page provides further information for handling
+   vmx images built with KIWI and references the following
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
   vmx images with encrypted root partitions built with KIWI and references the following
```

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   articles:
+
+   * :ref:`vmx`
+
+A virtual disk image can be partially or fully encrypted
+using the LUKS extension supported by KIWI. A fully encrypted
+image also includes the data in :file:`/boot` to be encrypted.
+Such an image requests the passphrase for the master key
+to be entered at the bootloader stage. A partialy encrypted
+image keeps :file:`/boot` open and on an extra boot partition.
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
image keeps :file:`/boot` unencrypted and on an extra boot partition.
```

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+Such an image requests the passphrase for the master key later
+in the boot process when the root partition gets accessed by
+the systemd mount service. In any case the master passphrase
+is requested only once.
+
+Update the KIWI image description as follows:
+
+1. Software packages
+
+   Make sure to add the following packages to the package list
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
   Make sure to add the following package to the package list:
```

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+   .. note::
+
+      Package names used in the following list matches the
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
      Package names used in the following list match the
```

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+   .. code:: xml
+
+      <package name="cryptsetup"/>
+
+2. Image Type definition
+
+   Update the vmx image type setup as follows
+
+   Full Encryption Including :file:`/boot`
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
   Full disk encryption including :file:`/boot`:
```

File: doc/source/building/working_with_images/vmx_setup_for_luks.rst [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+        <type image="vmx"
+            image="vmx"
+            filesystem="ext4"
+            bootloader="grub2"
+            luks="linux"
+            bootpartition="false">
+        </type>
+
+   Partial Encryption with :file:`/boot` on an extra partition
```
Comment (dcermak) [inline doc/source/building/working_with_images/vmx_setup_for_luks.rst]:
```suggestion
   Encrypted :file:`/` with an unencrypted :file:`/boot` partition:
```

**Comment by schaefi** [comment]
@davidcassany I was reading the cryptomount code a bit and also looked for the docs:

* https://www.gnu.org/software/grub/manual/grub/html_node/cryptomount.html

So far there is no way to provide the master key passphrase as a commandline argument. This means the process here is interactive. It would be super cool if one can pass in the passphrase on the grub level in another way than terminal input. We could then provide encrypted images also for the cloud. But I think so far there is no way of doing it.

Also just realized that the EFI boot is not complete and missing the cryptomount setup in the early boot script. Will add that in the next hour

**Pull request merged**